cpp/src/arrow/type.cc

```diff
@@ -1459,6 +1478,42 @@ int32_t DecimalType::DecimalSize(int32_t precision) {
   return static_cast<int32_t>(std::ceil((precision / 8.0) * std::log2(10) + 1));
 }
 
+// ----------------------------------------------------------------------
+// Decimal32 type
+
+Decimal32Type::Decimal32Type(int32_t precision, int32_t scale)
+    : DecimalType(type_id, 4, precision, scale) {
+  ARROW_CHECK_GE(precision, kMinPrecision);
+  ARROW_CHECK_LE(precision, kMaxPrecision);
```

Comment:
Would it be worthwhile to extract

```
template <typename D>
Status ValidateDecimalPrecision(int32_t precision) {
  if (precision < D::kMinPrecision || precision > D::kMaxPrecision) {
    return Status::Invalid("Decimal precision out of range [", int32_t(D::kMinPrecision),
                           ", ", int32_t(D::kMaxPrecision), "]: ", precision);
  }
  return Status::Ok();
}
```
?

Then we could write
```suggestion
  ARROW_CHECK_OK(ValidatePrecisionScale<Decimal32Type>());
```

And in the `::Make` functions below
```
Result<std::shared_ptr<DataType>> Decimal32Type::Make(int32_t precision, int32_t scale) {
  RETURN_NOT_OK(ValidatePrecisionScale<Decimal32Type>(precision));
  return std::make_shared<Decimal32Type>(precision, scale);
}
```

Reply:
made these changes